File: src/main/java/cz/dynawest/csvcruncher/HsqlDbHelper.java
package cz.dynawest.csvcruncher;

import cz.dynawest.csvcruncher.util.DbUtils;
import cz.dynawest.csvcruncher.util.Utils;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.SQLSyntaxErrorException;
import java.sql.Statement;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;

@Slf4j
public class HsqlDbHelper
{
    private static final Logger LOG = log;
    public static final int MAX_STRING_COLUMN_LENGTH = 4092;


    private Connection jdbcConn;

    public HsqlDbHelper(Connection jdbcConn)
    {
        this.jdbcConn = jdbcConn;
    }


    void createTableForInputFile(String tableName, File csvFileToBind, List<String> colNames, boolean ignoreFirst, boolean overwrite) throws SQLException
    {
        createTableAndBindCsv(tableName, csvFileToBind, colNames, ignoreFirst, "", true, overwrite);
    }

    private void createTableAndBindCsv(String tableName, File csvFileToBind, List<String> columnsNames, boolean ignoreFirst, String counterColumnDdl, boolean isInputTable, boolean overwrite) throws SQLException
    {
        Map<String, String> columnsDef = listToMapKeysWithNullValues(columnsNames);
        createTableAndBindCsv(tableName, csvFileToBind, columnsDef, ignoreFirst, counterColumnDdl, isInputTable, overwrite);

        // Try to convert columns types to numbers, where applicable.
        if (isInputTable) {
            this.optimizeTableCoumnsType(tableName, columnsNames);
        }
    }

    /**
     * Returns a map with keys from the given list, and null values. Doesn't deal with duplicate keys.
     */
    private static Map<String, String> listToMapKeysWithNullValues(List<String> keys)
    {
        LinkedHashMap<String, String> result = new LinkedHashMap<>();
        for (String columnsName : keys) {
            result.put(columnsName, null);
        }
        return result;
    }

    /**
     * Creates the input or output table, with the right column names, and binds the file.<br/>
     * For output tables, the file is optionally overwritten if exists.<br/>
     * A header with columns names is added to the output table.<br/>
     * Input tables columns are optimized after binding the file by attempting to reduce the column type.
     * (The output table has to be optimized later.)<br/>
     */
    void createTableAndBindCsv(String tableName, File csvFileToBind, Map<String, String> columnsNamesAndTypes, boolean ignoreFirst, String counterColumnDdl, boolean isInputTable, boolean overwrite) throws SQLException
    {
        boolean readOnly = false;
        boolean csvUsesSingleQuote = true;

        if (isInputTable && !csvFileToBind.exists()) {
            throw new IllegalArgumentException("The input file does not exist: " + csvFileToBind);
        }

        // Get a full path, because HSQLDB resolves paths against the data dir specified in JDBC URL.
        try {
            csvFileToBind = csvFileToBind.getCanonicalFile();
        }
        catch (IOException ex) {
            throw new CsvCruncherException("Failed resolving the CSV file path: " + csvFileToBind, ex);
        }

        // Delete any file at the output path, if exists. Other option would be to TRUNCATE, but this is safer.
        if (!isInputTable) {
            if (csvFileToBind.exists()) {
                if (true || overwrite) // TODO: Obey --overwrite.
                    csvFileToBind.delete();
                else
                    throw new IllegalArgumentException("The output file already exists. Use --overwrite or delete: " + csvFileToBind);
            }
            else {
                try {
                    Files.createDirectories(csvFileToBind.getParentFile().toPath());
                }
                catch (IOException ex) {
                    throw new CsvCruncherException("Failed creating directory to store the output to: " + csvFileToBind.getParentFile(), ex);
                }
            }
        }



        // We are also building a header for the CSV file.
        StringBuilder sbCsvHeader = new StringBuilder("# ");
        StringBuilder sbSql = (new StringBuilder("CREATE TEXT TABLE ")).append(tableName).append(" ( ");

        // The counter column, if any.
        sbSql.append(counterColumnDdl);

        // Columns
        for (Map.Entry<String, String> columnDef : columnsNamesAndTypes.entrySet())
        {
            String columnName = Utils.escapeSql(columnDef.getKey());
            String columnType = columnDef.getValue();
            if (columnType == null || "VARCHAR".equals(columnType.toUpperCase()))
                columnType = "VARCHAR(" + MAX_STRING_COLUMN_LENGTH + ")";
            else
                columnType = Utils.escapeSql(columnType);

            sbCsvHeader.append(columnName).append(", ");
            sbSql.append(columnName).append(" ").append(columnType).append(", ");
        }
        sbCsvHeader.delete(sbCsvHeader.length() - 2, sbCsvHeader.length());
        sbSql.delete(sbSql.length() - 2, sbSql.length());
        sbSql.append(" )");
        LOG.debug("Table DDL SQL: " + sbSql.toString());
        this.executeDbCommand(sbSql.toString(), "Failed to CREATE TEXT TABLE: ");


        // Bind the table to the CSV file.
        String csvPath = csvFileToBind.getPath();
        csvPath = Utils.escapeSql(csvPath);
        String quoteCharacter = csvUsesSingleQuote ? "\\quote" : "\"";
        String ignoreFirstFlag = ignoreFirst ? "ignore_first=true;" : "";
        String csvSettings = "encoding=UTF-8;cache_rows=50000;cache_size=10240000;" + ignoreFirstFlag + "fs=,;qc=" + quoteCharacter;
        String DESC = readOnly ? "DESC" : "";  // Not a mistake, HSQLDB really has "DESC" here for read only.
        String sql = String.format("SET TABLE %s SOURCE '%s;%s' %s", tableName, csvPath, csvSettings, DESC);
        LOG.debug("CSV import SQL: " + sql);
        this.executeDbCommand(sql, "Failed to import CSV: ");

        // SET TABLE <table name> SOURCE HEADER
        if (!isInputTable) {
            sql = String.format("SET TABLE %s SOURCE HEADER '%s'", tableName, sbCsvHeader.toString());
            LOG.debug("CSV source header SQL: " + sql);
            this.executeDbCommand(sql, "Failed to set CSV header: ");
        }

    }


    /**
     * Execute a SQL which does not expect a ResultSet,
     * and help the user with the common errors by parsing the message
     * and printing out some helpful info in a wrapping exception.
     *
     * @return the number of affected rows.
     */
    public int executeDbCommand(String sql, String errorMsg)
    {
        try (Statement stmt = this.jdbcConn.createStatement()){
            return stmt.executeUpdate(sql);
        }
        catch (Exception ex) {
            String addToMsg = "";
            //if (ex.getMessage().contains("for cast"))
            {
                // List column names with types.
                addToMsg = "\n  Tables and column types:\n"
                        + this.formatListOfAvailableTables(true);
            }

            if (ex.getMessage().contains("cannot be converted to target type")) {
                errorMsg = StringUtils.defaultString(errorMsg) + " Looks like the data in the input files do not match.";
            }

            if (StringUtils.isBlank(errorMsg))
                errorMsg = "Looks like there was a data type mismatch. Check the output table column types and your SQL.";

            throw new CsvCruncherException(errorMsg
                    + "\n  SQL: " + sql
                    + "\n  DB error: " + ex.getClass().getSimpleName() + " " + ex.getMessage()
                    + addToMsg
            );
        }
    }

    /**
     * Prepares a list of tables in the given JDBC connections, in the PUBLIC schema.
     */
    public String formatListOfAvailableTables(boolean withColumns)
    {
        String schema = "'PUBLIC'";

        StringBuilder sb = new StringBuilder();
        String sqlTablesMetadata =
                "SELECT table_name AS t, c.column_name AS c, c.data_type AS ct" +
                        " FROM INFORMATION_SCHEMA.TABLES AS t " +
                        " NATURAL JOIN INFORMATION_SCHEMA.COLUMNS AS c " +
                        " WHERE t.table_schema = " + schema;

        try (Statement st = jdbcConn.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_READ_ONLY)) {
            ResultSet rs = st.executeQuery(sqlTablesMetadata);

            tables:
            while(rs.next()) {
                String tableName = rs.getString("T");
                sb.append(" * ").append(tableName).append('\n');
                while(StringUtils.equals(tableName, rs.getString("T"))) {
                    if (withColumns)
                        sb.append("    - ")
                                .append(StringUtils.rightPad(rs.getString("C"), 28))
                                .append(" ")
                                .append(rs.getString("CT"))
                                .append('\n');
                    if (!rs.next())
                        break tables;
                }
                rs.previous();
            }
            if (sb.length() == 0)
                return "    (No tables)";
            return sb.toString();
        }
        catch (SQLException ex) {
            String msg = "Failed listing tables: " + ex.getMessage();
            LOG.error(msg);
            return msg;
        }
    }

    /**
     * Analyzes the exception against the given DB connection and rethrows an exception with a message containing the available objects as a hint.
     */
    public CsvCruncherException throwHintForObjectNotFound(SQLSyntaxErrorException ex)
    {
        boolean notFoundIsColumn = DbUtils.analyzeWhatWasNotFound(ex.getMessage());

        String tableNames = formatListOfAvailableTables(notFoundIsColumn);

        String hintMsg = notFoundIsColumn ?
                "\n  Looks like you are referring to a column that is not present in the table(s).\n"
                        + "  Check the header (first line) in the CSV.\n"
                        + "  Here are the tables and columns are actually available:\n"
                :
                "\n  Looks like you are referring to a table that was not created.\n"
                        + "  This could mean that you have a typo in the input file name,\n"
                        + "  or maybe you use --combineInputs but try to use the original inputs.\n"
                        + "  These tables are actually available:\n";

        return new CsvCruncherException(
                hintMsg
                        + tableNames + "\nMessage from the database:\n  "
                        + ex.getMessage(), ex);
    }

    /**
     * Get the columns info: Perform the SQL, LIMIT 1.
     */
    public Map<String, String> extractColumnsInfoFrom1LineSelect(String sql) throws SQLException
    {
        PreparedStatement statement;
        try {
            statement = jdbcConn.prepareStatement(sql + " LIMIT 1");
        }
        catch (SQLSyntaxErrorException ex) {
            if (ex.getMessage().contains("object not found:")) {
                throw throwHintForObjectNotFound(ex);
            }
            throw new CsvCruncherException("Seems your SQL contains errors:\n" + ex.getMessage(), ex);
        }
        catch (SQLException ex) {
            throw new CsvCruncherException("Failed executing the SQL:\n" + ex.getMessage(), ex);
        }
        ResultSet rs = statement.executeQuery();

        // Column names
        return DbUtils.getResultSetColumnNamesAndTypes(rs);
    }


    /**
     * Detaches or re-attaches HSQLDB TEXT table.
     * @param drop     Drop the table after detaching.
     */
    void detachTable(String tableName, boolean drop) throws SQLException
    {
        LOG.debug(String.format("Detaching%s table: %s", drop ? " and dropping" : "", tableName));

        String sql = "SET TABLE " + Utils.escapeSql(tableName) + " SOURCE OFF";
        executeDbCommand(sql, "Failed to detach/attach the table: ");

        if (drop) {
            sql = "DROP TABLE " + Utils.escapeSql(tableName);
            executeDbCommand(sql, "Failed to DROP TABLE: ");
        }
    }

    void attachTable(String tableName) throws SQLException
    {
        LOG.debug("Ataching table: " + tableName);

        String sql = "SET TABLE " + Utils.escapeSql(tableName) + " SOURCE ON";
        executeDbCommand(sql, "Failed to attach the table: ");
    }

    /**
     * This must be called when all data are already in the table!
     * Try to convert columns to best fitting types.
     * This speeds up further SQL operations.
     *
     *   "HyperSQL allows changing the type if all the existing values can be cast
     *    into the new type without string truncation or loss of significant digits."
     */
    void optimizeTableCoumnsType(String tableName, List<String> colNames) throws SQLException
    {
        Map<String, String> columnsFitIntoType = new LinkedHashMap<>();

        // TODO: This doesn't work because: Operation is not allowed on text table with data in statement.
        // See https://stackoverflow.com/questions/52647738/hsqldb-hypersql-changing-column-type-in-a-text-table
        // Maybe I need to duplicate the TEXT table into a native table first?
        for (String colName : colNames)
        {
            String typeUsed;
            for (String sqlType : new String[]{"VARCHAR(255)", "CHAR", "TIMESTAMP", "UUID", "BIGINT", "INTEGER", "SMALLINT", "BOOLEAN"})
            {
                // Try CAST( AS ...)
                String sqlCol = String.format("SELECT CAST(%s AS %s) FROM %s", colName, sqlType, tableName);
                //String sqlCol = String.format("SELECT 1 + \"%s\" FROM %s", colName, tableName);

                LOG.trace("Column change attempt SQL: " + sqlCol);
                try (Statement st = jdbcConn.createStatement()) {
                    st.execute(sqlCol);
                    LOG.trace(String.format("Column %s.%s fits to %s", tableName, colName, typeUsed = sqlType));
                    columnsFitIntoType.put(colName, sqlType);
                }
                catch (SQLException ex) {
                    // LOG.trace(String.format("Column %s.%s values don't fit to %s.\n  %s", tableName, colName, sqlType, ex.getMessage()));
                }
            }
        }

        detachTable(tableName, false);

        // ALTER COLUMNs
        for (Map.Entry<String, String> colNameAndType : columnsFitIntoType.entrySet())
        {
            String colName = colNameAndType.getKey();
            String sqlType = colNameAndType.getValue();
            String sqlAlter = String.format("ALTER TABLE %s ALTER COLUMN %s SET DATA TYPE %s", tableName, colName, sqlType);
            String sqlCheck = String.format("SELECT data_type FROM information_schema.columns WHERE LOWER(table_name) = LOWER('%s') AND LOWER(column_name) = LOWER('%s')", tableName, colName);

            LOG.trace("Changing the column {} to {}", colName, sqlType);
            try (Statement st = jdbcConn.createStatement()) {
                st.execute(sqlAlter);
                LOG.info(String.format("Column %s.%s converted to %s. %s", tableName, colName, sqlType, sqlAlter));
                LOG.trace("Checking col type: " + sqlCheck);
                ResultSet columnTypeRes = st.executeQuery(sqlCheck);
                if (!columnTypeRes.next()) {
                    LOG.error("Column not found?? {}.{}", tableName, colName);
                    DbUtils.testDumpSelect("SELECT table_name, column_name FROM information_schema.columns WHERE LOWER(table_name) = LOWER('"+tableName.toUpperCase()+"')", jdbcConn);
                    continue;
                }
                String newType = columnTypeRes.getString("data_type");
                if (!newType.equals(sqlType) ) {
                    LOG.error(String.format("Column %s.%s did not really change the type to %s, stayed %s.", tableName, colName, sqlType, newType));
                }
            }
            catch (SQLException ex) {
                LOG.error(String.format("Error changing type of column %s.%s to %s.\n  %s", tableName, colName, sqlType, ex.getMessage()));
            }
        }

        attachTable(tableName);
    }

    public void detachTables(Set<String> tableNames, String msgOnError)
    {
        for (String tableName : tableNames) {
            try {
                detachTable(tableName, true);
            } catch (Exception ex) {
                LOG.error(msgOnError + ex.getMessage());
            }
        }

    }


    static String normalizeFileNameForTableName(File fileName)
    {
        return fileName.getName().replaceFirst(".csv$", "").replaceAll("[^a-zA-Z0-9_]", "_");
    }

}

File: src/main/kotlin/cz/dynawest/csvcruncher/HsqlDbHelper.kt
package cz.dynawest.csvcruncher

import cz.dynawest.csvcruncher.util.DbUtils.getResultSetColumnLabelsAndTypes
import cz.dynawest.csvcruncher.util.HsqldbErrorHandling.throwHintForObjectNotFound
import cz.dynawest.csvcruncher.util.logger
import org.apache.commons.lang3.StringUtils
import java.nio.file.Path
import java.sql.*

@Suppress("NAME_SHADOWING")
class HsqlDbHelper(val jdbcConn: Connection) {

    /**
     * Execute a SQL which does not expect a ResultSet,
     * and help the user with the common errors by parsing the message
     * and printing out some helpful info in a wrapping exception.
     *
     * @return the number of affected rows.
     */
    fun executeSql(sql: String, errorMsg: String): Int {
        if (sql.isBlank()) throw CsvCruncherException("$errorMsg: SQL is an empty string.")

        var errorMsg = errorMsg
        try {
            log.debug("    Executing SQL: $sql")
            jdbcConn.createStatement().use { stmt -> return stmt.executeUpdate(sql) }
        }
        // TBD: Handle SQLSyntaxErrorException specifically
        catch (ex: Exception) {
            var addToMsg = ""
            if (true || (ex.message?.contains("for cast") ?: false)) {
                // List column names with types.
                addToMsg = """
                    |
                    |  Tables and column types:
                    |${this.formatListOfAvailableTables(true)}""".trimMargin()
            }
            if (ex.message!!.contains("cannot be converted to target type")) {
                errorMsg = "$errorMsg Looks like the data in the input files do not match."
            }
            if (errorMsg.isBlank()) errorMsg = "Looks like there was a data type mismatch. Check the output table column types and your SQL."
            throw CsvCruncherException("""$errorMsg
                |  DB error: ${ex.javaClass.simpleName} ${ex.message}$addToMsg
                |  SQL: $sql
                """.trimMargin()
            )
        }
    }

    /**
     * Prepares a list of tables in the given JDBC connections, in the PUBLIC schema.
     */
    fun formatListOfAvailableTables(withColumns: Boolean): String {
        val schema = "PUBLIC"
        val sqlTablesMetadata =
            "SELECT t.table_name AS t, c.column_name AS c, c.data_type AS ct " +
            " FROM INFORMATION_SCHEMA.TABLES AS t" +
            " NATURAL JOIN INFORMATION_SCHEMA.COLUMNS AS c  WHERE t.table_schema = '$schema'"

        try {
            jdbcConn.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_READ_ONLY).use { st ->
                val rs = st.executeQuery(sqlTablesMetadata)
                val sb = StringBuilder()
                tables@ while (rs.next()) {
                    val tableName = rs.getString("T")
                    sb.append(" * ").append(tableName).append('\n')
                    while (StringUtils.equals(tableName, rs.getString("T"))) {
                        if (withColumns) sb.append("    - ")
                                .append(StringUtils.rightPad(rs.getString("C"), 28))
                                .append(" ")
                                .append(rs.getString("CT"))
                                .append('\n')
                        if (!rs.next()) break@tables
                    }
                    rs.previous()
                }
                return if (sb.isEmpty()) "    (No tables)" else sb.toString()
            }
        } catch (ex: SQLException) {
            val msg = "Failed listing tables: " + ex.message
            log.error(msg)
            return msg
        }
    }


    /**
     * Get the columns info: Perform the SQL, LIMIT 1.
     */
    @Throws(SQLException::class)
    fun extractColumnsInfoFrom1LineSelect(sql: String): Map<String, String> {

        val sql = setOrReplaceLimit(sql, "LIMIT 1")

        val statement: PreparedStatement
        statement = try {
            jdbcConn.prepareStatement(sql)
        }
        catch (ex: SQLSyntaxErrorException) {
            if (ex.message!!.contains("unexpected token:")) {
                throw SqlSyntaxCruncherException("""
                |    The SQL contains syntax error:
                |    ${ex.message}
                |    $sql
                |    This may be your SQL error or caused by alteration by CsvCruncher. 
                |    See https://github.com/OndraZizka/csv-cruncher/issues for known bugs.
                |    """.trimMargin())
            }
            if (ex.message!!.contains("object not found:")) {
                throw throwHintForObjectNotFound(ex, this, sql)
            }
            throw CsvCruncherException("""
                |    Seems your SQL contains errors:
                |    ${ex.message}
                |    $sql
                |    """.trimMargin(), ex)
        } catch (ex: SQLException) {
            throw CsvCruncherException("""
                |    Failed executing the SQL:
                |    ${ex.message}
                |    $sql
                |    """.trimMargin(), ex)
        }
        val rs = statement.executeQuery()

        // Column names
        return getResultSetColumnLabelsAndTypes(rs)
    }


    /**
     * Detaches or re-attaches HSQLDB TEXT table.
     * @param drop     Drop the table after detaching.
     */
    @Throws(SQLException::class)
    fun detachTable(tableName: String, drop: Boolean) {
        log.debug("Detaching${if (drop) " and dropping" else ""} table: $tableName")
        var sql = "SET TABLE ${quote(tableName)} SOURCE OFF"
        executeSql(sql, "Failed to detach/attach the table: ")
        if (drop) {
            sql = "DROP TABLE ${quote(tableName)}"
            executeSql(sql, "Failed to DROP TABLE: ")
        }
    }

    @Throws(SQLException::class)
    fun attachTable(tableName: String) {
        log.debug("Attaching table: $tableName")
        val sql = "SET TABLE ${quote(tableName)} SOURCE ON"
        executeSql(sql, "Failed to attach the table: ")
    }



    class ColumnNotFoundException(tableName: String, colName: String?) : Exception("Column not found: $tableName.$colName")

    fun detachTables(tableNames: Set<String>, msgOnError: String) {
        for (tableName in tableNames) {
            try {
                detachTable(tableName, true)
            } catch (ex: Exception) {
                log.warn(msgOnError + ex.message)
            }
        }
    }

    fun queryAllTableAndColumnNames(includeTables: Boolean = false, includeColumns: Boolean = true): Set<String> {
        val schema = "PUBLIC"
        val sqlTablesMetadata =
            """SELECT t.table_name AS "tableName", c.column_name AS "colName" 
                 FROM INFORMATION_SCHEMA.TABLES AS t
                 NATURAL JOIN INFORMATION_SCHEMA.COLUMNS AS c  
                 WHERE t.table_schema = '$schema'
             """

        try {
            val identifiers = mutableSetOf<String>()
            jdbcConn.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE, ResultSet.CONCUR_READ_ONLY).use { st ->
                val resultSet = st.executeQuery(sqlTablesMetadata)
                while (resultSet.next()) {
                    if (includeTables)  identifiers.add(resultSet.getString("tableName"))
                    if (includeColumns) identifiers.add(resultSet.getString("colName"))
                }
            }
            return identifiers
        } catch (ex: SQLException) {
            throw CsvCruncherException("Couldn't list all columns: ${ex.message}", ex)
        }
    }

    fun executeSqlScript(path: Path, errorMsg: String) {
        if (!path.toFile().exists()) throw CsvCruncherException("$errorMsg: Does not exist: $path")
        if (!path.toFile().isFile()) throw CsvCruncherException("$errorMsg: Is not a file: $path")

        for (line in path.toFile().readLines().asSequence().map { it.trim().removeSuffix(";") }) {
            if (line.isBlank()) continue
            if (line.startsWith("--")) continue
            executeSql(line, errorMsg)
        }
    }

    fun quoteColumnAndTableNamesInQuery(sqlQuery: String): String {
        val identifiers = mutableListOf<String>()
        identifiers.addAll(queryAllTableAndColumnNames(includeTables = true, includeColumns = true))
        return quoteIdentifiersInQuery(sqlQuery, identifiers)
    }


    companion object {
        private val log = logger()
        const val MAX_STRING_COLUMN_LENGTH = 4092

        fun normalizeTableName(proposedName: String): String {
            return proposedName.replace("[^a-zA-Z0-9_]".toRegex(), "_")
        }

        fun quote(identifier: String) = "\"${identifier.replace('"', '\'')}\""

        fun setOrReplaceLimit(sql: String, newLimitClause: String) = sql.replace("\\s*\\bLIMIT\\s+\\d+(\\s+OFFSET\\s+\\d+)?\\s*\$".toRegex(), "") + " $newLimitClause"

        /**
         * Because of HSQLDB's rules of column names normalization, the column names need to be quoted in queries (or be all uppercased).
         * To relieve the user from quoting everything, this method does it.
         * Although, the current impl is potentially brittle.
         */
        fun quoteIdentifiersInQuery(sqlQuery: String, identifiers: List<String>): String {
            var sqlQuery = sqlQuery
            val columnNames = identifiers.sortedByDescending { it.length }
            val substitutes = mutableMapOf<String, String>()

            for (name in columnNames.withIndex()) {
                val substitute = "{.${name.index}}"
                substitutes[substitute] = "\"${name.value}\""
                val escapedName = Regex.escape(name.value)
                sqlQuery = sqlQuery.replace("""(?i)\b(?<!")$escapedName(?!")\b""".toRegex(), Regex.escapeReplacement(substitute))
            }
            for (i in columnNames.indices) {
                val substitute = "{.${i}}"
                sqlQuery = sqlQuery.replace(substitute, substitutes.get(substitute)!!)
            }

            return sqlQuery
        }
    }
}
